### PR TITLE
test(integration): run integration test for TiDB

### PR DESCRIPTION
**Pull request by Mini256**
### What does this PR do?

In this PR, I will replicate the MySQL integration test case to test TiDB. 

part of #14639

### About TiDB

> [TiDB](https://github.com/pingcap/tidb) (/’taɪdiːbi:/, "Ti" stands for Titanium) is an open-source NewSQL database that supports Hybrid Transactional and Analytical Processing (HTAP) workloads. It is MySQL compatible and features horizontal scalability, strong consistency, and high availability. The goal of TiDB is to provide users with a one-stop database solution that covers OLTP (Online Transactional Processing), OLAP (Online Analytical Processing), and HTAP services. TiDB is suitable for various use cases that require high availability and strong consistency with large-scale data.

- [GitHub Repo: pingcap/tidb](https://github.com/pingcap/tidb) (with 34.2k🌟 and 5.5k🍴)
- [TiDB Cloud](https://en.pingcap.com/tidb-cloud/) (provide serverless cluster for developers)
- [Document about MySQL Compatibility](https://docs.pingcap.com/tidb/stable/mysql-compatibility)

### Test steps

1. Build the prisma binary

```shell
pnpm i
pnpm run build
```

2. Create the tidb testcases via copying the mysql testcase

```shell
cd packages/integration-tests
cp -r ./src/__tests__/integration/mysql ./src/__tests__/integration/tidb
```

3. Start up a TiDB cluster via `tiup playground`

```shell
# Install the deployment tool TiUP.
curl --proto '=https' --tlsv1.2 -sSf https://tiup-mirrors.pingcap.com/install.sh | sh
# Start a minimum cluster locally.
tiup playground v7.0.0 --without-monitor --tiflash 0
```

4. Run the testes:

```shell
export TEST_TIDB_BASE_URI="mysql://root@127.0.0.1:4000"
pnpm run test:tidb
```

### Found issues

In the first running of test, 4 test cases failed:

1. introspection: findUnique where PK with include (The naming strategy of constraint and the index generated by the foreign key is incompatible with MySQL. To fix it, modify the jest snapshot, for example: the default constraint name in MySQL is `posts_ibfk_1`, but in TiDB is `fk_1`)
2. introspection: findUnique where unique with foreign key and unpack (ditto)
3. introspection: findUnique where composite PK with foreign key (ditto)
4. runtime: findMany where in[string] (If the SQL statement does not declare order by, TiDB does not guarantee the order of the returned results, so to fix it, add `orderBy: { job: 'asc' }` for the testcase)

<details>

<summary>The first running result</summary>

```
➜  integration-tests git:(test-for-tidb) ✗ pnpm run test:tidb                                   

> @prisma/integration-tests@0.0.0 test:tidb /Users/liangzhiyuan/Projects/prisma/packages/integration-tests
> jest integration.tidb --maxConcurrency=1 --silent

 FAIL  src/__tests__/integration/tidb/runtime.test.ts
  ● Test suite failed to run

    TypeError: Do not know how to serialize a BigInt
        at stringify (<anonymous>)

      at messageParent (../../node_modules/.pnpm/jest-worker@29.5.0/node_modules/jest-worker/build/workers/messageParent.js:29:19)

 FAIL  src/__tests__/integration/tidb/introspection.test.ts (23.615 s)
  ● introspection: findUnique where PK with include

    expect(received).toMatchSnapshot(hint)

    Snapshot name: `introspection: findUnique where PK with include: datamodel 1`

    - Snapshot  - 2
    + Received  + 2

    @@ -10,13 +10,13 @@

      model posts {
        id      BigInt @id @default(autoincrement()) @mysql.UnsignedBigInt
        user_id BigInt @mysql.UnsignedBigInt
        title   String @mysql.VarChar(50)
    -   users   users  @relation(fields: [user_id], references: [id], onDelete: NoAction, map: "posts_ibfk_1")
    +   users   users  @relation(fields: [user_id], references: [id], onDelete: NoAction, map: "fk_1")

    -   @@index([user_id], map: "user_id")
    +   @@index([user_id], map: "fk_1")
      }

      model users {
        id    BigInt  @id @default(autoincrement()) @mysql.UnsignedBigInt
        email String  @unique(map: "email") @mysql.VarChar(50)

      201 |       states[scenario.name] = state
      202 |
    > 203 |       expect(introspectionResult.datamodel).toMatchSnapshot(`datamodel`)
          |                                             ^
      204 |       expect(introspectionResult.warnings).toMatchSnapshot(`warnings`)
      205 |
      206 |       await teardownScenario(state)

      at toMatchSnapshot (src/__tests__/__helpers__/integrationTest.ts:203:45)

  ● introspection: findUnique where unique with foreign key and unpack

    expect(received).toMatchSnapshot(hint)

    Snapshot name: `introspection: findUnique where unique with foreign key and unpack: datamodel 1`

    - Snapshot  - 2
    + Received  + 2

    @@ -10,13 +10,13 @@

      model posts {
        id      BigInt @id @default(autoincrement()) @mysql.UnsignedBigInt
        user_id BigInt @mysql.UnsignedBigInt
        title   String @mysql.VarChar(50)
    -   users   users  @relation(fields: [user_id], references: [id], onDelete: NoAction, map: "posts_ibfk_1")
    +   users   users  @relation(fields: [user_id], references: [id], onDelete: NoAction, map: "fk_1")

    -   @@index([user_id], map: "user_id")
    +   @@index([user_id], map: "fk_1")
      }

      model users {
        id    BigInt  @id @default(autoincrement()) @mysql.UnsignedBigInt
        email String  @unique(map: "email") @mysql.VarChar(50)

      201 |       states[scenario.name] = state
      202 |
    > 203 |       expect(introspectionResult.datamodel).toMatchSnapshot(`datamodel`)
          |                                             ^
      204 |       expect(introspectionResult.warnings).toMatchSnapshot(`warnings`)
      205 |
      206 |       await teardownScenario(state)

      at toMatchSnapshot (src/__tests__/__helpers__/integrationTest.ts:203:45)

  ● introspection: findUnique where composite PK with foreign key

    expect(received).toMatchSnapshot(hint)

    Snapshot name: `introspection: findUnique where composite PK with foreign key: datamodel 1`

    - Snapshot  - 2
    + Received  + 2

    @@ -18,10 +18,10 @@

      model b {
        id  BigInt @id @default(autoincrement()) @mysql.UnsignedBigInt
        one Int
        two Int
    -   a   a      @relation(fields: [one, two], references: [one, two], onDelete: NoAction, onUpdate: NoAction, map: "b_ibfk_1")
    +   a   a      @relation(fields: [one, two], references: [one, two], onDelete: NoAction, onUpdate: NoAction, map: "fk_1")

    -   @@index([one, two], map: "one")
    +   @@index([one, two], map: "fk_1")
      }
      ↵

      201 |       states[scenario.name] = state
      202 |
    > 203 |       expect(introspectionResult.datamodel).toMatchSnapshot(`datamodel`)
          |                                             ^
      204 |       expect(introspectionResult.warnings).toMatchSnapshot(`warnings`)
      205 |
      206 |       await teardownScenario(state)

      at toMatchSnapshot (src/__tests__/__helpers__/integrationTest.ts:203:45)

 › 3 snapshots failed.
Snapshot Summary
 › 3 snapshots failed from 1 test suite. Inspect your code changes or run `npm run test:tidb -- -u` to update them.

Test Suites: 2 failed, 2 total
Tests:       3 failed, 62 passed, 65 total
Snapshots:   3 failed, 127 passed, 130 total
Time:        23.809 s
```

</details>

Finally, after some minor changes mentioned above, all test cases can be passed by TiDB.

### CODE REVIEW

**Comment by Mini256** [comment]
If it is possible, I would like to add this set of tests to GitHub Action.

**Comment by Jolg42** [comment]
Hi @Mini256 

Thanks for your interest in adding tests for a new database (TiDB)! I'm also curious about it, though let me explain the situation about new databases:

Prisma is currently focusing on the existing supported database providers ([PostgreSQL, MySQL, MariaDB, CockroachDB, SQLite & MongoDB](https://www.prisma.io/docs/reference/database-reference/supported-databases)), and we do not plan to add or put effort to support new database at the moment.

About adding support to a new database, from our experience, it requires a lot of effort and often additional code even when a database claims to be "fully wire compatible with X".
The tests you added here are only a small step, and it would require many changes to fully tests all Prisma parts here in prisma/prisma monorepo and https://github.com/prisma/prisma-engines for example.

It would be helpful for us if you could open a [feature request issue here](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) about that you would want Prisma to support TiDB so we can collect all the interest from different people in one place. Feel free to add context about this, it really helps to understand why this would be important, for example.

**Comment by Jolg42** [comment]
Note @Mini256 

So, this would only be merged if we decide that 
- from a product perspective, it makes sense
- tests in this git monorepo pass
- tests in prisma-engines also fully pass

If, for example, tests would all pass, and this is essentially the same as a currently supported provider like MySQL/MariaDB, that might change our opinion, and we might be much more open to talk about it for getting a product decision 👍🏼

**Comment by janpio** [comment]
Thanks for your PR @Mini256, exciting. The most important thing that Joel wrote is our wish for a feature request issue for TiDB support. That is where everything starts in our process.

Still, this PR is great as it just shows what is working, and what is not.

To give you some more input:

> Because TiDB currently does not support foreign keys (which is working in process: [pingcap/tidb#18209](https://github.com/pingcap/tidb/issues/18209)), test cases `1.`, `2.` and `3.` cannot pass. I set them to `todo: true`, once the foreign keys feature is supported, I will reopen these use cases.

We actually already support PlanetScale/Vitess that also does not support foreign keys (you can read more about this here: https://www.prisma.io/docs/guides/database/using-prisma-with-planetscale), but are not explicitly testing that here in this repository. In `prisma-engines`, our other repository, we have a full test suite running against Vitess to make sure we cover all the queries and specialties of that database. That might be your next step to see what works, and what does not for TiDB. (The testing setup is a bit more complex, so feel free to ask about details if you can not figure it out).

**Comment by ghimsim** [comment]
Hello Jan and Joel,

Thank you for your detailed responses. We have filed a feature request to support TiDB to Prisma: https://github.com/prisma/prisma/issues/14639

I totally understand your need to prioritize your support for databases. However, we are an open source database trying to get traction in the marketplace. We need support from great software like Prisma to get going, so we have somewhat of a chicken and egg thing here - to get more customers we need Prisma, but to get into Prisma, we need more customers.

If we are willing to put in engineering resources to help with the tests, integration and anything else, would you be more open to adding TiDB to your list of supported databases?

Ghim

**Comment by janpio** [comment]
Thanks @ghimsim for that context.

Disclaimer: I can only speak for the Engineering org side of Prisma, not the Product one (but their concerns would probably be similar).

As you certainly know, even if you help us develop this initially, there is a long tail of follow on costs that Prisma will have to carry. Random examples: Longer test runtimes, additional complexity in developing new features, additional things to keep in mind when debugging or making changes. Those are all real costs that **can not be prevented** even if you promise to help us with this "forever".

So currently this PR and discussion is mostly figuring out the current state of Prisma + TiDB:

1. If TiDB indeed turns out to be equivalent to Vitess/PlanetScale (so a quirky variant of MySQL), then we would post that in the feature request issue and could then have an internal discussion with Prisma if we are _officially_ willing to carry declare that and carry TiDB as a supported database. 
   - (If we decide not to, this would still allow _you_ to claim inofficial support by Prisma via `mysql` and `referentialIntegrity=prisma` - which would basically be the same thing for your users.)
3. If it is not, that means either TiDB or Prisma needs to change in some way to _get_ to that state anyway. 
   - In coordination with us you could certainly contribute changes to Prisma, that could bring us closer - but of course we would need to check if we take on additional responsibility because of that. There might be things we do not want to accept. 
   - We could also help you to understand what TiDB needs to change or adapt to be able to pass our tests.
   - Maybe this leads to us creating additional feature request issues (similar to our preview feature `referentialintegrity=prisma`) for things that could be useful for all databases. That would then again need prioritization from us, but bring us closer to support that functionality both for TiDB and others.

So for now, figuring out exactly what is happening here and seeing what happens in `prisma-engines` seems like good next steps to me (now that the feature request issue was created).

File: packages/integration-tests/src/__tests__/integration/tidb/__database.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+      password: credentials.password,
+      multipleStatements: true,
+    })
+  },
+  beforeEach: async (db, sqlScenario, ctx) => {
+    const sqlUp = `
+    DROP DATABASE IF EXISTS ${ctx.id};
+    CREATE DATABASE ${ctx.id};
+    USE ${ctx.id};`
+    await db.query(sqlUp + sqlScenario)
```
[the remark this answers is not in the document]
Comment (Mini256) [reply packages/integration-tests/src/__tests__/integration/tidb/__database.ts:26]: @sonatype-lift ignore

**Comment by janpio** [comment]
@Mini256 took this suggestion, and https://github.com/prisma/prisma-engines/pull/3308 now exists <3

File: .buildkite/test/docker-compose.14.yml [hunk 2]
```diff
@@ -125,6 +129,12 @@ services:
     ports:
       - '4306:3306'
 
+  tidb:
+    image: pingcap/tidb:v6.1.0
```
Comment (Mini256) [inline .buildkite/test/docker-compose.14.yml:153]:
```suggestion
    image: pingcap/tidb:v7.0.0
```

**Review by Mini256 — COMMENTED** [review]
update image to v7.0.0

**Comment by Jolg42** [comment]
Note: I noticed tests were not running and fixed that in https://github.com/prisma/prisma/commit/82a11547c5d01a372809ee1b511d39645847d043

**Comment by laplab** [comment]
Thank you for submitting this PR, @Mini256!

We mentioned in [our documentation](https://www.prisma.io/docs/orm/overview/databases/database-drivers#community-maintained-database-driver-adapters) that TiDB has a community-maintained [driver adapter](https://github.com/tidbcloud/prisma-adapter). This driver adapter already allows users to effortlessly query TiDB databases. Thus, we currently do not plan to merge this PR and https://github.com/prisma/prisma-engines/pull/3308.

**Pull request closed without merging**